cli/src/debugger.rs

```diff
@@ -0,0 +1,857 @@
+            .create(flags.contains(HostIoOpenFlags::O_CREAT))
+            .append(flags.contains(HostIoOpenFlags::O_APPEND))
+            .truncate(flags.contains(HostIoOpenFlags::O_TRUNC));
+        let _ = mode; // TODO
+        let file = match options.open::<&str>(path.as_ref()) {
+            Ok(file) => file,
+            Err(err) => {
+                return Err(HostIoError::from(err));
+            }
+        };
+        // TODO windows
+        use std::os::fd::AsRawFd;
+        let fd = file.as_raw_fd() as u32;
+        match self.open_files.insert(fd, file) {
+            Some(_) => panic!("open: file descriptor already in use"),
+            None => Ok(fd),
+        }
+    }
+}
+
+impl HostIoClose for MachineTarget {
+    fn close(&mut self, fd: u32) -> HostIoResult<(), Self> {
+        match self.open_files.remove(&fd) {
+            Some(file) => file.sync_all().map_err(|e| HostIoError::from(e)),
+            None => Err(HostIoError::from(std::io::Error::new(
+                ErrorKind::NotFound,
+                "close: file descriptor not found",
+            ))),
+        }
+    }
+}
+
+impl HostIoPread for MachineTarget {
+    fn pread(
+        &mut self,
+        fd: u32,
+        count: usize,
+        offset: u64,
+        buf: &mut [u8],
+    ) -> HostIoResult<usize, Self> {
+        // TODO windows
+        use std::os::unix::fs::FileExt;
+        let Some(file) = self.open_files.get_mut(&fd) else {
+            return Err(HostIoError::from(std::io::Error::new(
+                ErrorKind::NotFound,
+                "pread: file descriptor not found",
+            )));
+        };
+        let end = count.min(buf.len());
+        file.read_at(&mut buf[..end], offset)
+            .map_err(|e| HostIoError::from(e))
+    }
+}
+
+impl SingleThreadResume for MachineTarget {
+    fn resume(&mut self, signal: Option<Signal>) -> Result<(), Self::Error> {
+        let _ = signal;
+        self.action = Some(MachineTargetAction::Resume);
+        Ok(())
+    }
+
+    fn support_single_step(&mut self) -> Option<SingleThreadSingleStepOps<'_, Self>> {
+        Some(self)
+    }
+
+    fn support_range_step(&mut self) -> Option<SingleThreadRangeSteppingOps<'_, Self>> {
+        None // TODO
+    }
+}
+
+impl SingleThreadSingleStep for MachineTarget {
+    fn step(&mut self, signal: Option<Signal>) -> Result<(), Self::Error> {
+        let _ = signal;
+        self.action = Some(MachineTargetAction::SingleStep);
+        Ok(())
+    }
+}
+
+fn wait_for_gdb_connection(port: u16) -> std::io::Result<std::net::TcpStream> {
+    let sockaddr = format!("localhost:{}", port);
+    log::info!("Waiting for a GDB connection on {:?}...", sockaddr);
+    let sock = std::net::TcpListener::bind(sockaddr)?;
+    let (stream, addr) = sock.accept()?;
+    log::info!("Debugger connected from {}", addr);
+    Ok(stream)
+}
+
+pub type StateMachine<'a> = GdbStubStateMachine<'a, MachineTarget, std::net::TcpStream>;
```

Comment:
I am trying to follow why you went with implementing this StateMachine type rather that using the simpler event loop from the gdbstub examples. Can you explain a bit?

I would have expected something more like this
https://github.com/daniel5151/gdbstub/blob/6e72c26211515bf15b8a462d195f6ccf418f4c92/examples/armv4t/main.rs#L95